File: Dynamic-Programming/1143. Longest Common Subsequence/longestCommonSubsequence.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
// bottom up approach
// Time Complexity: O(m * n)
var longestCommonSubsequence = function (text1, text2) {
    // init number of rows 
    let dp = new Array(text1.length + 1).fill(0);
    // init each row with value
    for (let i = 0; i < dp.length; i++) {
        dp[i] = new Array(text2.length + 1).fill(0)
    }

    for (let i = text1.length - 1; i >= 0; i--) {
        for (let j = text2.length - 1; j >= 0; j--) {
            if (text1[i] === text2[j]) {
                // common character - diagnomally 
                dp[i][j] = 1 + dp[i + 1][j + 1];
            } else {
                dp[i][j] = Math.max(dp[i][j + 1], dp[i + 1][j]);
            }
        }
    }

    // [
    // [ 3, 2, 1, 0 ],
    // [ 2, 2, 1, 0 ],
    // [ 2, 2, 1, 0 ],
    // [ 1, 1, 1, 0 ],
    // [ 1, 1, 1, 0 ],
    // [ 0, 0, 0, 0 ]
    // ]
    return dp[0][0];
};

File: Structures & Algo/D. Dynamic Programming/Longest Common Subsequence/1143. Longest Common Subsequence/longestCommonSubsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        @lru_cache(None)
        def dp(i, j):
            if i == -1 or j == -1:
                return 0
            
            if text1[i] == text2[j]:
                return dp(i - 1, j - 1) + 1
            else:
                return max(dp(i - 1, j), dp(i, j - 1))
        
        return dp(len(text1) - 1, len(text2) - 1)

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m = len(text1) 
        n = len(text2)
        
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if text1[i - 1] == text2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + 1
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
        
        # return dp[m][n]
        return dp[-1][-1]